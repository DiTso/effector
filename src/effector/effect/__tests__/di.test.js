//@flow

import {createEffect} from '..'
import type {FnEffect} from '../index.h'

const delay = n => new Promise(rs => setTimeout(rs, n))

test('effect.create single argument', async() => {
  const effect = createEffect('long request')
  effect.use(async() => {
    await delay(500)
    return 'done!'
  })
  const oldCreate = effect.create
  //$off
  effect.create = jest.fn((payload, fullName, args) =>
    oldCreate(payload, fullName, args),
  )
  const baz = jest.fn()
  effect.done.watch(baz)

  await effect(100)
  await effect(200)
  await effect(300)

  await expect(effect.create.mock.calls).toMatchSnapshot()
  await expect(baz.mock.calls).toMatchSnapshot()
})

//eslint-disable-next-line
type ExtractFn<Params, Fn: (...params: Params) => any> = Params

function variadicEffect<Done, Fn:(...params: any[]) => Promise<Done> | Done>(
  name?: string,
): FnEffect<ExtractFn<*, Fn>, Done, Error, Fn> {
  const effect = createEffect(name)
  const oldCreate = effect.create
  //$off
  effect.create = jest.fn((payload, fullName, args) =>
    oldCreate([payload, ...args], fullName, []),
  )
  const oldUse = effect.use
  //$off
  effect.use = thunk => oldUse(payload => thunk(...payload))

  return (effect: any)
}

test('effect.create multiple arguments', async() => {
  const useSpy = jest.fn()
  const baz = jest.fn()
  const effect = variadicEffect<
    string,
    *,
    (a: number, b: number) => Promise<string>,
      >('long request')

  effect.use(async(a, b) => {
    useSpy(a, b)
    await delay(500)
    return 'done!'
  })
  effect.done.watch(baz)

  await effect(100, 200)
  await effect(200, 300)
  await effect(300, 400)

  await expect(effect.create.mock.calls).toMatchSnapshot('effect.create')
  await expect(useSpy.mock.calls).toMatchSnapshot('effect.use')
  await expect(baz.mock.calls).toMatchSnapshot('effect.watch')
})
