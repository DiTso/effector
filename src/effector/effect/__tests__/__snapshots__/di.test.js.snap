// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`effect.create multiple arguments: effect.create 1`] = `
Array [
  Array [
    100,
    "long request",
    Array [
      200,
    ],
  ],
  Array [
    200,
    "long request",
    Array [
      300,
    ],
  ],
  Array [
    300,
    "long request",
    Array [
      400,
    ],
  ],
]
`;

exports[`effect.create multiple arguments: effect.use 1`] = `
Array [
  Array [
    100,
    200,
  ],
  Array [
    200,
    300,
  ],
  Array [
    300,
    400,
  ],
]
`;

exports[`effect.create multiple arguments: effect.watch 1`] = `
Array [
  Array [
    Object {
      "params": Array [
        100,
        200,
      ],
      "result": "done!",
    },
  ],
  Array [
    Object {
      "params": Array [
        200,
        300,
      ],
      "result": "done!",
    },
  ],
  Array [
    Object {
      "params": Array [
        300,
        400,
      ],
      "result": "done!",
    },
  ],
]
`;

exports[`effect.create single argument 1`] = `
Array [
  Array [
    100,
    "long request",
    Array [],
  ],
  Array [
    200,
    "long request",
    Array [],
  ],
  Array [
    300,
    "long request",
    Array [],
  ],
]
`;

exports[`effect.create single argument 2`] = `
Array [
  Array [
    Object {
      "params": 100,
      "result": "done!",
    },
  ],
  Array [
    Object {
      "params": 200,
      "result": "done!",
    },
  ],
  Array [
    Object {
      "params": 300,
      "result": "done!",
    },
  ],
]
`;
